src/sciline/serialize/_json.py

```diff
@@ -0,0 +1,146 @@
+from typing import Union
+
+from .._provider import Provider
+from .._utils import key_full_qualname, key_name, provider_full_qualname, provider_name
+from ..typing import Graph, Json, Key
+
+
+def json_schema() -> Json:
+    """Return the JSON schema for serialized task graphs.
+
+    See, e.g., `jsonschema <https://python-jsonschema.readthedocs.io/en/stable/>`_
+    for a tool that can validate the schema.
+
+    Returns
+    -------
+    :
+        The graph JSON schema as a dict.
+    """
+    with (
+        importlib.resources.files('sciline.serialize')
+        .joinpath('graph_json_schema.json')
+        .open()
+    ) as f:
+        return json.load(f)  # type: ignore[no-any-return]
+
+
+def json_serialize_task_graph(graph: Graph) -> dict[str, Json]:
+    """Serialize a graph to JSON.
+
+    See the user guide on
+    `Serializing Providers <../../user-guide/serialization.rst>`_.
+
+    Also available as :meth:`sciline.TaskGraph.serialize`.
+
+    Returns
+    -------
+    :
+        A JSON object representing the graph.
+    """
+    id_gen = _IdGenerator()
+    nodes = []
+    edges = []
+    for key, provider in graph.items():
+        n, e = _serialize_provider(key, provider, id_gen)
+        nodes.append(n)
+        edges.extend(e)
+    return {
+        'directed': True,
+        'multigraph': False,
+        'nodes': nodes,  # type: ignore[dict-item]
+        'edges': edges,  # type: ignore[dict-item]
+    }
+
+
+def _serialize_provider(
+    key: Key, provider: Provider, id_gen: _IdGenerator
+) -> tuple[dict[str, Json], list[dict[str, Json]]]:
+    if provider.kind == 'function':
+        return _serialize_function(key, provider, id_gen)
+    if provider.kind == 'parameter':
+        return _serialize_param(key, id_gen)
+    raise ValueError(
+        f'Cannot serialize a task graph that contains {provider.kind} nodes.'
+    )
+
+
+def _serialize_param(
+    key: Key, id_gen: _IdGenerator
+) -> tuple[dict[str, Json], list[dict[str, Json]]]:
+    node = {
+        'id': id_gen.node_id(key),
+        'kind': 'parameter',
+        'label': key_name(key),
+        'out': key_full_qualname(key),
+    }
+    return node, []
+
+
+def _serialize_function(
+    key: Key, provider: Provider, id_gen: _IdGenerator
+) -> tuple[dict[str, Json], list[dict[str, Json]]]:
+    node_id = id_gen.node_id(key)
+
+    edges = []
+    args = []
+    kwargs = {}
+    for i, arg in enumerate(provider.arg_spec.args):
+        edge = _serialize_edge(arg, key, i, id_gen)
+        edges.append(edge)
+        args.append(edge['id'])
+    for name, kwarg in provider.arg_spec.kwargs:
+        edge = _serialize_edge(kwarg, key, name, id_gen)
+        edges.append(edge)
+        kwargs[name] = edge['id']
+
+    node = {
+        'id': node_id,
+        'kind': 'function',
+        'label': provider_name(provider),
+        'function': provider_full_qualname(provider),
+        'out': key_full_qualname(key),
+        'args': args,
+        'kwargs': kwargs,
+    }
```

Comment:
Without having looked at everything, I think this looks much better now (with the Sciline-specifics removed) 👍 

Question: With the `out` as part of the node, we make all dependent nodes linked to the `id` of the provider. Maybe that is not so great, and we should have an `id` for the data, i.e., re-introduce "data" nodes so they can have their own `id`? Or do you think it does not matter?

Reply:
I had a long discussion with Max about this. Separating data and functions is in principle cleaner. It would even remove the need for the parameter kind. And if we ever need to attach attributes to data, we could do so without affecting function nodes.

However, it makes the graph harder to read and the whole process a little more complicated. So the thought was to go with what I implemented now and see how that goes and change later if need be. We still have a little time to figure things out.

Reply:
> Separating data and functions is in principle cleaner.

I thought it would be cleaner especially since we also have input nodes already, a special kind of data node.

> However, it makes the graph harder to read

By whom?

> and the whole process a little more complicated.

Wouldn't it simply mean that providers need to add two nodes, the compute node and the output data node?

Reply:
> I thought it would be cleaner especially since we also have input nodes already, a special kind of data node.

This is what I meant by 

> > It would even remove the need for the parameter kind

> > However, it makes the graph harder to read
>
> By whom?
>
> > and the whole process a little more complicated.
>
> Wouldn't it simply mean that providers need to add two nodes, the compute node and the output data node?

By humans because they need to follow two edges to find how data relates to each other instead of one. And by computers because they need to take into account the two fundamentally different node types. E.g., for reconstructing a task graph, the reader must extract pairs of nodes for each task instead of just one.

For writing, this is not that much more complicated and well contained.

Reply:
> But they are connected through a direct edge? Is it actually complicated enough to warrant combining compute and data nodes?

No. This makes a miniscule difference on all points unless we foresee adding more fields to data nodes.
I'm happy to change back. I mainly went this way because you objected to having separate notes initially.

Reply:
> I mainly went this way because you objected to having separate notes initially.

I did not object, I mainly asked about it! And my main concern there was the discrepancy between inputs (then stored as functions) and data nodes.

It seems Aiida has data nodes? https://aiida.readthedocs.io/projects/aiida-core/en/latest/howto/visualising_graphs.html

Reply:
Yes, they do. This links into what I hinted at with storing extra metadata in data nodes. If we, e.g., stored intermediate results in a database or in files, we could link to them here, kind of like Aiida does.